frame/uniques/src/lib.rs

```diff
@@ -0,0 +1,1182 @@
+			let class_details = Class::<T, I>::get(&class).ok_or(Error::<T, I>::Unknown)?;
+			ensure!(class_details.admin == origin, Error::<T, I>::NoPermission);
+
+			let source = details.owner;
+			details.owner = dest;
+
+			Account::<T, I>::remove((&source, &class, &instance));
+			Account::<T, I>::insert((&details.owner, &class, &instance), ());
+
+			Asset::<T, I>::insert(&class, &instance, &details);
+
+			Self::deposit_event(Event::Transferred(class, instance, source, details.owner));
+
+			Ok(())
+		}
+
+		/// Disallow further unprivileged transfer of an asset instance.
+		///
+		/// Origin must be Signed and the sender should be the Freezer of the asset `class`.
+		///
+		/// - `class`: The class of the asset to be frozen.
+		/// - `instance`: The instance of the asset to be frozen.
+		///
+		/// Emits `Frozen`.
+		///
+		/// Weight: `O(1)`
+		#[pallet::weight(T::WeightInfo::freeze())]
+		pub(super) fn freeze(
+			origin: OriginFor<T>,
+			#[pallet::compact] class: T::ClassId,
+			#[pallet::compact] instance: T::InstanceId,
+		) -> DispatchResult {
+			let origin = ensure_signed(origin)?;
+
+			let mut details = Asset::<T, I>::get(&class, &instance)
+				.ok_or(Error::<T, I>::Unknown)?;
+			let class_details = Class::<T, I>::get(&class).ok_or(Error::<T, I>::Unknown)?;
+			ensure!(class_details.freezer == origin, Error::<T, I>::NoPermission);
+
+			details.is_frozen = true;
+			Asset::<T, I>::insert(&class, &instance, &details);
+
+			Self::deposit_event(Event::<T, I>::Frozen(class, instance));
+			Ok(())
+		}
+
+		/// Re-allow unprivileged transfer of an asset instance.
+		///
+		/// Origin must be Signed and the sender should be the Freezer of the asset `class`.
+		///
+		/// - `class`: The class of the asset to be thawed.
+		/// - `instance`: The instance of the asset to be thawed.
+		///
+		/// Emits `Thawed`.
+		///
+		/// Weight: `O(1)`
+		#[pallet::weight(T::WeightInfo::thaw())]
+		pub(super) fn thaw(
+			origin: OriginFor<T>,
+			#[pallet::compact] class: T::ClassId,
+			#[pallet::compact] instance: T::InstanceId,
+		) -> DispatchResult {
+			let origin = ensure_signed(origin)?;
+
+			let mut details = Asset::<T, I>::get(&class, &instance)
+				.ok_or(Error::<T, I>::Unknown)?;
+			let class_details = Class::<T, I>::get(&class).ok_or(Error::<T, I>::Unknown)?;
+			ensure!(class_details.admin == origin, Error::<T, I>::NoPermission);
+
+			details.is_frozen = false;
+			Asset::<T, I>::insert(&class, &instance, &details);
+
+			Self::deposit_event(Event::<T, I>::Frozen(class, instance));
+			Ok(())
+		}
+
+		/// Disallow further unprivileged transfers for a whole asset class.
+		///
+		/// Origin must be Signed and the sender should be the Freezer of the asset `class`.
+		///
+		/// - `class`: The asset class to be frozen.
+		///
+		/// Emits `ClassFrozen`.
+		///
+		/// Weight: `O(1)`
+		#[pallet::weight(T::WeightInfo::freeze_class())]
+		pub(super) fn freeze_class(
```

Comment:
`freeze_user` may also be a helpful utility

Reply:
Would require an additional storage item and an extra read for every transfer operation. Overall would complicate the storage model quite badly and probably introduce account reference requirements. Just a huge headache.